### ISSUE TITLE
rpc: read-your-writes consistency on `/tx` endpoint

### ISSUE THREAD

**Issue by tarcieri** [tendermint/tendermint#7196]
I had originally filed this as #6359, which includes ample relevant context.

**Environment**:

@jackzampolin's gaia test container: https://hub.docker.com/r/jackzampolin/gaiatest

**What happened**:

After submitting a valid `MsgSend` transaction via the JSON RPC API using the `/broadcast_tx_commit` endpoint and receiving successful response for both `CheckTx` and `DeliverTx`, I attempted to look up the commit hash in the response using the `/tx` RPC endpoint.

I observed there is a small window, ostensibly between when the transaction is accepted and when it is indexed, where `/tx` will return no results:

https://github.com/cosmos/cosmos-rust/pull/116/checks?check_run_id=3724891908

```
Caused by:
    Internal error: tx (FFE0D601EF72615F73F9C3F99F583A1D1DBB2C9A010EB910185E88D2048E1E7C) not found (code: -32603)
```

**What you expected to happen**:

When performing a `/tx` request after receiving confirmation of a successfully accepted transaction,  I was expecting a successful response with the transaction.

In the context of an integration test I was writing, I was hoping for [read-your-writes (RYW) consistency](http://www.dbms2.com/2010/05/01/ryw-read-your-writes-consistency/), namely that performing a `/tx` request for a transaction with successful `DeliverTx` is guaranteed to successfully return the executed transaction in a response.

**Comment by tychoish** [comment]
This is super tricky! We've been talking a lot internally about the `broadcast_tx_commit` endpoint, (which I think was added originally for testing, and has for better or worse become part of the public API.) Historically, and in 0.34 and before, the implementation used a single blocking subscription to inform users when their transaction was committed, which would be susceptible to the bug described here (if the subscription for the RPC call was processed before the subscription for the indexer.)

In 0.35, this changes in this commit https://github.com/tendermint/tendermint/commit/b203c9179916b12f1068358ba7b13d4b7393f722, which means we poll the indexer (if enabled) after calling `CheckTx` on the transaction and return successful only when the transaction has been committed. I believe that this means that the new method would provide RYW semantics relative to `/tx` end point. The subscription/events system is structured in such a way that a number of blocking subscriptions could put backpressure on core areas of the system, and cause the node to fall behind in block production or consensus: the new implementation not only provides better RYW semantics, but also removes the liability of the event system in this RPC call. 

In either implementation it's possible for the method to return an error for a transaction that has successfully entered the mempool, and would eventually be committed, because of a baseline timeout within the RPC method implementation, and I think the correct solution would be to poll the `/tx` endpoint client-side with your own timeout/backoff logic. Under the hood, in the new implementation of the method, this is exactly what's happening, though it is subject to the RPC systems baseline timeout.

**Comment by tarcieri** [comment]
>  We've been talking a lot internally about the broadcast_tx_commit endpoint, (which I think was added originally for testing

I'm also dealing with this in a testing use case. I'm okay with the solution is "proper patterns for testing" rather than first-class features.

>  I think the correct solution would be to poll the /tx endpoint client-side with your own timeout/backoff logic. Under the hood, in the new implementation of the method, this is exactly what's happening, though it is subject to the RPC systems baseline timeout.

That's what I'm doing (polling), and it's working reliably for our test cases.

It seems like together with the application of `timeout_height` polling would suffice to establish whether or not a transaction was successfully committed to the chain or not, although that's not a use case of immediate concern to me.

**Comment by tychoish** [comment]
I’d be interested to learn if this issue is resolved for you in 0.35. I suspect it will be, though I think your polling solution for testing is probably superior even so. 

Do you think that there is more work we should or could do on this?

**Comment by tarcieri** [comment]
If there's a reliable way to make RYW work using polling + timeout, where after a timeout it's known that the transaction won't be accepted in the future, I guess that's fine.
